core/drivers/stm32_rng.c

```diff
@@ -428,7 +483,42 @@ static TEE_Result stm32_rng_pm_resume(uint32_t pm_cr)
 		if (io_read32(base + RNG_CR) & RNG_CR_CONDRST)
 			panic();
 	} else {
-		io_write32(base + RNG_CR, RNG_CR_RNGEN | pm_cr);
+		io_write32(base + RNG_CR, RNG_CR_RNGEN | stm32_rng->pm_cr);
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32_rng_pm_suspend(void)
+{
+	vaddr_t rng_base = get_base();
+
+	stm32_rng->pm_cr = io_read32(rng_base + RNG_CR);
+
+	if (stm32_rng->ddata->has_cond_reset) {
+		stm32_rng->pm_health = io_read32(rng_base + RNG_HTCR);
+		stm32_rng->pm_noise_ctrl = io_read32(rng_base + RNG_NSCR);
+	}
+
+	if (stm32_rng->ddata->has_power_optim) {
+		uint64_t timeout_ref = 0;
+
+		/*
+		 * As per reference manual, it is recommended to set
+		 * RNG_CONFIG2[bit0] when RNG power consumption is critical.
+		 */
+		io_setbits32(rng_base + RNG_CR, RNG_CR_POWER_OPTIM |
+				RNG_CR_CONDRST);
+		io_clrbits32(rng_base + RNG_CR, RNG_CR_CONDRST);
+
+		timeout_ref = timeout_init_us(RNG_READY_TIMEOUT_US);
+		while (io_read32(rng_base + RNG_CR) & RNG_CR_CONDRST)
+			if (timeout_elapsed(timeout_ref))
+				break;
+		if (io_read32(rng_base + RNG_CR) & RNG_CR_CONDRST)
+			panic();
```

Comment:
for info, could use:
```patch
+		uint32_t value = 0;

(...)

-		timeout_ref = timeout_init_us(RNG_READY_TIMEOUT_US);
-		while (io_read32(rng_base + RNG_CR) & RNG_CR_CONDRST)
-			if (timeout_elapsed(timeout_ref))
-				break;
-		if (io_read32(rng_base + RNG_CR) & RNG_CR_CONDRST)
+		if (IO_READ32_POLL_TIMEOUT(rng_base + RNG_CR, value,
+					   !(value & RNG_CR_CONDRST), 0,
+					   RNG_READY_TIMEOUT_US))
 			panic();

Reply:
I'll change to this implementation, I'd rather use these kinds of macros when possible